fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,244 @@
+#![allow(dead_code, unused)] // FIXME: stop shouting rustc
+
+use std::{collections::BTreeMap, future::pending, sync::Arc, time::Duration};
+
+use anyhow::{bail, Context as _};
+use async_trait::async_trait;
+use fedimint_core::{
+    apply, async_trait_maybe_send,
+    config::{ClientConfig, GlobalClientConfig},
+    db::{Committable, Database, DatabaseTransaction, IDatabaseTransactionOpsCoreTyped},
+    task::{timeout, MaybeSend, MaybeSync},
+    util::{retry, FibonacciBackoff},
+};
+use futures::{future::join_all, StreamExt};
+use serde::de::DeserializeOwned;
+use tracing::{debug, instrument, warn};
+
+use crate::{
+    db::{MetaFieldKey, MetaFieldName},
+    Client,
+};
+
+/// Tries to parse `str_value` as JSON. In the special case that `V` is `String`
+/// we return the raw `str_value` if JSON parsing fails. This necessary since
+/// the spec wasn't clear enough in the beginning.
+#[instrument(err)] // log on every failure
+pub fn parse_meta_value_static<V: DeserializeOwned + 'static>(
+    str_value: &str,
+) -> anyhow::Result<V> {
+    let res = serde_json::from_str(str_value)
+        .with_context(|| format!("Decoding meta field value '{str_value}' failed"));
+
+    // In the past we encoded some string fields as "just a string" without quotes,
+    // this code ensures that old meta values still parse since config is hard to
+    // change
+    if res.is_err() && std::any::TypeId::of::<V>() == std::any::TypeId::of::<String>() {
+        let string_ret = Box::new(str_value.to_owned());
+        let ret: Box<V> = unsafe {
+            // We can transmute a String to V because we know that V==String
+            std::mem::transmute(string_ret)
+        };
+        Ok(*ret)
+    } else {
+        res
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+pub trait MetaSource: MaybeSend + MaybeSync {
+    /// Higher value overwrites sources with lower priority value.
+    fn priority(&self) -> u64;
+    /// Wait for next change in this source.
+    async fn wait_for_update(&self);
+    async fn fetch(&self, client: &Client) -> anyhow::Result<BTreeMap<String, String>>;
+}
+
+#[derive(Default, Clone)]
+pub struct MetaManager {
+    sources: Arc<Vec<Box<dyn MetaSource>>>,
+}
+
+impl MetaManager {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub async fn get_field<V: DeserializeOwned + 'static>(
+        &self,
+        db: &Database,
+        field: &str,
+    ) -> Option<V> {
+        // TODO: maybe wait for initial sources
+        db.begin_transaction_nc()
+            .await
+            .find_by_prefix_sorted_descending(&MetaFieldName(field.to_string()))
+            .await
+            .collect::<Vec<_>>()
+            .await
+            .into_iter()
+            .find_map(|(_, value)| parse_meta_value_static::<V>(&value).ok())
+    }
+
+    pub(crate) fn add_source(&mut self, source: impl MetaSource + 'static) {
+        Arc::get_mut(&mut self.sources)
+            .expect("add_source on non unique MetaManager")
```

Comment:
This cannot happen because `add_source` is only used before the client is constructed, thus before one could access the `MetaManager` from outside?

Reply:
moved this expect to client builder to make logic more local